### PR TITLE
Provide layer specific fns on `ServiceBuilder`

### PR DESCRIPTION
**Pull request by carllerche**
This allows avoiding the layer type imports.

### CODE REVIEW

File: tower/src/builder/mod.rs [hunk 1]
```diff
@@ -172,6 +182,41 @@ impl<L> ServiceBuilder<L> {
         }
     }
 
+    /// Buffer requests
+    pub fn buffer(self, bound: usize) -> ServiceBuilder<Chain<BufferLayer, L>> {
+        self.layer(BufferLayer::new(bound))
+    }
+
+    /// Filter requests
+    pub fn filter<U>(self, predicate: U) -> ServiceBuilder<Chain<FilterLayer<U>, L>> {
```
Comment (LucioFranco) [inline tower/src/builder/mod.rs]: should we have a where clause here to bound what `U` should be to not require the user to go to `FilterLayer::new` to find the bound?

File: tower/src/builder/mod.rs [hunk 2]
```diff
@@ -172,6 +182,41 @@ impl<L> ServiceBuilder<L> {
...
+        self.layer(LoadShedLayer::new())
+    }
+
+    /// Limit requests to at most `num` per the given duration
+    pub fn rate_limit(self, num: u64, per: Duration) -> ServiceBuilder<Chain<RateLimitLayer, L>> {
+        self.layer(RateLimitLayer::new(num, per))
+    }
+
+    /// Retry failed requests.
+    pub fn retry<P>(self, policy: P) -> ServiceBuilder<Chain<RetryLayer<P>, L>> {
```
Comment (LucioFranco) [inline tower/src/builder/mod.rs]: same here we should bound this by Policy

File: tower/src/builder/mod.rs (hunk 1 above)
Comment (carllerche) [reply tower/src/builder/mod.rs]: We cannot add bounds because we don't know the type of `Request`.

**Comment by carllerche** [comment]
I added some documentation covering ordering and the implicit trait bounds.

**Pull request merged**